x/mongo/driver/topology/server_test.go

```diff
@@ -54,6 +54,298 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+				}
+			}()
+
+			// Create a TCP listener on a random port. The listener will accept connections but not
+			// read or write to them.
+			l, err := net.Listen("tcp", "127.0.0.1:0")
+			require.NoError(t, err)
+			defer func() {
+				_ = l.Close()
+			}()
+
+			server, err := NewServer(
+				address.Address(l.Addr().String()),
+				primitive.NewObjectID(),
+				// Create a connection pool event monitor that sends all events to an events channel
+				// so we can assert on the connection pool events later.
+				WithConnectionPoolMonitor(func(_ *event.PoolMonitor) *event.PoolMonitor {
+					return &event.PoolMonitor{
+						Event: func(event *event.PoolEvent) {
+							eventsCh <- event
+						},
+					}
+				}),
+				// Replace the default dialer and handshaker with the test dialer and handshaker, if
+				// present.
+				WithConnectionOptions(func(opts ...ConnectionOption) []ConnectionOption {
+					if tc.connectTimeout > 0 {
+						opts = append(opts, WithConnectTimeout(func(time.Duration) time.Duration { return tc.connectTimeout }))
+					}
+					if tc.dialer != nil {
+						opts = append(opts, WithDialer(tc.dialer))
+					}
+					if tc.handshaker != nil {
+						opts = append(opts, WithHandshaker(tc.handshaker))
+					}
+					return opts
+				}),
+				// Disable monitoring to prevent unrelated failures from the RTT monitor and
+				// heartbeats from unexpectedly clearing the connection pool.
+				withMonitoringDisabled(func(bool) bool { return true }),
+			)
+			require.NoError(t, err)
+			require.NoError(t, server.Connect(nil))
+
+			// Create a context with the operation timeout if one is specified in the test case.
+			ctx := context.Background()
+			if tc.operationTimeout > 0 {
+				var cancel context.CancelFunc
+				ctx, cancel = context.WithTimeout(ctx, tc.operationTimeout)
+				defer cancel()
+			}
+			_, err = server.Connection(ctx)
+			if tc.expectErr {
+				assert.NotNil(t, err, "expected an error but got nil")
+			} else {
+				assert.Nil(t, err, "expected no error but got %s", err)
+			}
+
+			// Disconnect the server then close the events channel and expect that no more events
+			// are sent on the channel. Then wait for the events channel loop to return before
+			// inspecting the events slice.
+			_ = server.Disconnect(context.Background())
+			close(eventsCh)
+			eventsWg.Wait()
+			require.NotEmpty(t, events, "expected more than 0 connection pool monitor events")
+
+			// Look for any "ConnectionPoolCleared" events in the events slice so we can assert that
+			// the Server connection pool was or wasn't cleared, depending on the test expectations.
+			poolCleared := false
+			for _, evt := range events {
+				if evt.Type == event.PoolCleared {
+					poolCleared = true
+				}
+			}
+			assert.Equal(
+				t,
+				tc.expectPoolCleared,
+				poolCleared,
+				"want pool cleared: %t, actual pool cleared: %t",
+				tc.expectPoolCleared,
+				poolCleared)
+		})
+	}
+}
+
+func TestServerConnectionCancellation(t *testing.T) {
```

Comment:
nit: For readability, can you write a comment at the beginning of this test that explains what we're testing?

Reply:
Will add comment